spark/src/main/scala/org/apache/spark/sql/delta/schema/ImplicitMetadataOperation.scala

```diff
@@ -227,6 +227,149 @@ object ImplicitMetadataOperation {
+   * @param protocol the protocol used.
+   * @param metadata the metadata used for checking constraints and generated columns.
+   * @param currentDt the current data type.
+   * @param updateDt the updated data type.
+   */
+  private def checkConstraintsOrGeneratedColumnsExceptStruct(
+      spark: SparkSession,
+      path: Seq[String],
+      protocol: Protocol,
+      metadata: Metadata,
+      currentDt: DataType,
+      updateDt: DataType): Unit = (currentDt, updateDt) match {
+    case (StructType(_), StructType(_)) =>
+      // we explicitly ignore the check for `StructType` here.
+    case (from, to) =>
+      if (currentDt != updateDt) {
+        checkDependentConstraints(spark, path, metadata, from, to)
+        checkDependentGeneratedColumns(spark, path, protocol, metadata, from, to)
+      }
+  }
+
+  /**
+   * If the provided data types are `StructType`, check whether the nested/inner fields
+   * of the struct are being referenced by CHECK constraints or generated columns.
+   * It is worth noting that:
+   * 1. we only care about the inner fields of `StructType`,
+   *    for all the other types, e.g., `MapType` or `ArrayType`, we will directly check
+   *    the dependant constraints and generated columns without inspecting the inner fields;
+   *    this indicates we intentionally keep being too strict for `MapType` and `ArrayType`.
+   *
+   * 2. this function will not recursively check any nested `StructType`, the actual traverse
+   *    through the nested struct(s) will be handled by [[SchemaMergingUtils.transformColumns]]
+   *    in [[checkDependentExpressions]].
+   *
+   * @param spark the spark session used.
+   * @param protocol the protocol used.
+   * @param path the full column path for the current field.
+   * @param from the current data type.
+   * @param to the updated data type.
+   * @param metadata the metadata used for checking constraints and generated columns.
+   */
+  private def checkReferencedByCheckConstraintsOrGeneratedColumns(
+      spark: SparkSession,
+      protocol: Protocol,
+      path: Seq[String],
+      from: DataType,
+      to: DataType,
+      metadata: Metadata): Unit = (from, to) match {
+    case (StructType(fromFields), StructType(toFields)) =>
+      fromFields.zip(toFields).foreach { case (fromField, toField) =>
+        checkConstraintsOrGeneratedColumnsExceptStruct(
+          spark = spark,
+          path = path :+ fromField.name,
+          protocol = protocol,
+          metadata = metadata,
+          currentDt = fromField.dataType,
+          updateDt = toField.dataType
+        )
+      }
+    // do note that for `MapType` and `ArrayType`, we intentionally keep being too strict when
+    // checking nested fields - i.e., even if the containing nested fields are not
+    // being referenced by any dependant constraints or generated columns.
+    // reference: <https://github.com/delta-io/delta/pull/3601#discussion_r1731203107>.
```

Comment:
Personal opinion: I'm rather against putting links in comments, these can get out of sync, be inaccessible to some, and just generally require context switching to get information.
Might be better instead to give a quick explanation inline: the field paths for maps/arrays in constraints/generated expressions aren't consistent with regular field paths: `a.arr[0].x` vs. `a.element.x`. This makes it hard to recurse into arrays and maps here, so we prefer being too strict.

Reply:
agreed, I’ll replace the link with an explanatory comment.

thanks for the suggestion!